File: src/main/java/com/pandus/leetcode/solutions/google/FlipEquivalentBinaryTrees.java
package com.pandus.leetcode.solutions.google;

import com.pandus.leetcode.solutions.model.TreeNode;

public class FlipEquivalentBinaryTrees {
    public boolean flipEquiv(TreeNode root1, TreeNode root2) {
        return check(root1, root2);
    }

    private boolean check(TreeNode tree1, TreeNode tree2) {
        if (tree1 == null && tree2 == null) return true;
        if (tree1 == null || tree2 == null) return false;
        if (tree1.getVal() != tree2.getVal()) return false;

        return (check(tree1.getLeft(), tree2.getRight()) && check(tree1.getRight(), tree2.getLeft())
                || check(tree1.getLeft(), tree2.getLeft()) && check(tree1.getRight(), tree2.getRight()));
    }
}

File: src/main/kotlin/com/pandus/leetcode/solutions/daily/FlipEquivalentBinaryTrees.kt
package com.pandus.leetcode.solutions.daily

import com.pandus.leetcode.solutions.model.TreeNode

// Reference: https://leetcode.com/problems/flip-equivalent-binary-trees
class FlipEquivalentBinaryTrees {
    fun flipEquiv(root1: TreeNode?, root2: TreeNode?): Boolean {
        // Both trees are empty
        if (root1 == null && root2 == null) {
            return true
        }
        // Just one of the trees is empty
        if (root1 == null || root2 == null) {
            return false
        }
        // Corresponding values differ
        if (root1.`val` != root2.`val`) {
            return false
        }

        // Check if corresponding subtrees are flip equivalent
        val noSwap =
            flipEquiv(root1.left, root2.left) &&
                    flipEquiv(root1.right, root2.right)
        // Check if opposite subtrees are flip equivalent
        val swap =
            flipEquiv(root1.left, root2.right) &&
                    flipEquiv(root1.right, root2.left)

        return noSwap || swap
    }
}
